subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/SyncTaskIntegrationTest.groovy

```diff
@@ -394,36 +393,85 @@ class SyncTaskIntegrationTest extends AbstractIntegrationSpec {
-                }
+        buildFile << '''
+            task syncIt(type: Sync) {
+                from project.hasProperty("capitalizeFile") ? "FILE.TXT" : "file.txt"
+                from project.hasProperty("capitalizeDir") ? "DIR" : "dir"
+                into buildDir
             }
-            task sync(type: Sync) {
-                from writeFile
-                into new File(buildDir, "sync")
+        '''
+
+        def lowercaseFile = file('file.txt')
+        lowercaseFile.text = 'file'
+
+        file('dir').create {
+            file('file2.txt') << 'file2'
+            file('file3.txt') << 'file3'
+            file('file4.txt') << 'file4'
+            nestedDir {
+                file('nestedDirFile.txt') << 'nestedDirFile'
             }
-        """
+        }
+        and:
+        run 'syncIt'
+        file('build').assertHasDescendants(
+            'file.txt',
+            'file2.txt',
+            'file3.txt',
+            'file4.txt',
+            'nestedDir/nestedDirFile.txt'
+        )
+
+        and:
+        def uppercaseFile = new File(lowercaseFile.parentFile, 'FILE.TXT') // make sure case is kept
+        lowercaseFile.renameTo(uppercaseFile)
+        uppercaseFile.text = 'FILE'
+        assert uppercaseFile.canonicalFile.name == 'FILE.TXT'
+        and:
+        file('dir/file3.txt').delete()
+
+        file('dir/file4.txt').delete()
+        file('dir/File4.txt') << 'File4'
+
         when:
-        succeeds("sync")
+        succeeds('syncIt', '-PcapitalizeFile')
         then:
-        file("build/output").assertExists()
-        file("build/sync/output").assertExists()
+        executedAndNotSkipped ':syncIt'
+        file('build/FILE.TXT').with {
+            assert it.parentFile.list() != [].toArray()
+            assert it.assertExists()
+            assert it.text == 'FILE'
+            assert it.canonicalFile.name == 'FILE.TXT'
+        }
+        and:
+        file('build/file2.txt').assertExists()
+        file('build/file3.txt').assertDoesNotExist()
+        file('build/File4.txt').with {
+            it.assertExists()
+            it.canonicalFile.name == 'File4.txt'
+        }
 
         when:
-        succeeds("sync", "-Pcapitalize")
+        file('dir').deleteDir()
+        file('DIR').create {
+            file('file4.txt') << 'file4'
+            NESTEDDIR {
+                file('NESTEDDIRFILE.TXT') << 'NESTEDDIRFILE'
+            }
+        }
+        and:
+        succeeds('sync', '-PcapitalizeDir')
         then:
-        file("build/OUTPUT").assertExists()
-        // TODO: This should exist
-        file("build/sync/OUTPUT").assertDoesNotExist()
+        executedAndNotSkipped ':syncIt'
+        file('build/file4.txt').canonicalFile.name == 'file4.txt'
+        file('build/NESTEDDIR/NESTEDDIRFILE.TXT').with {
+            assert it.exists(): it.parentFile.list()
+            assert it.canonicalFile.name == 'NESTEDDIRFILE.TXT'
+            assert it.text == 'NESTEDDIRFILE'
```

Comment:
I think we also need to check that `build/NESTEDDIR` has the correct case.

Does this work now since you changed from delete to rename? When using delete, I would expect the method to fail since the directory is not empty. Actually, it would return `false`, right?

Reply:
Done.

Yes, I would need either `FileUtils.forceDelete` or `renameTo`.